src/util/cert/libcrypto/cert.c

```diff
@@ -176,31 +177,116 @@ errno_t sss_cert_pem_to_der(TALLOC_CTX *mem_ctx, const char *pem,
 #define IDENTIFIER_NISTP384 "nistp384"
 #define IDENTIFIER_NISTP521 "nistp521"
 
+static int sss_ec_get_key(BN_CTX *bn_ctx, EVP_PKEY *cert_pub_key,
+                          EC_GROUP **_ec_group, EC_POINT **_ec_public_key)
+{
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    int ret;
+    char curve_name[4096];
+    unsigned char pubkey[4096];
+    size_t len;
+    EC_GROUP *gr;
+    EC_POINT *pk;
+
+    ret = EVP_PKEY_get_utf8_string_param(cert_pub_key,
+                                         OSSL_PKEY_PARAM_GROUP_NAME,
+                                         curve_name, sizeof(curve_name), &len);
+    if (ret != 1) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    gr = EC_GROUP_new_by_curve_name(OBJ_sn2nid(curve_name));
+    if (gr == NULL) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = EVP_PKEY_get_octet_string_param(cert_pub_key,
+                                          OSSL_PKEY_PARAM_PUB_KEY,
+                                          pubkey, sizeof(pubkey), &len);
+    if (ret != 1) {
+        EC_GROUP_free(gr);
+        ret = EINVAL;
+        goto done;
+    }
+
+    pk = EC_POINT_new(gr);
+    if (pk == NULL) {
+        EC_GROUP_free(gr);
+        ret = EINVAL;
+        goto done;
+    }
+
+    ret = EC_POINT_oct2point(gr, pk, pubkey, len, bn_ctx);
+    if (ret != 1) {
+        EC_GROUP_free(gr);
+        EC_POINT_free(pk);
+        ret = EINVAL;
+        goto done;
+    }
+
+    *_ec_group = gr;
+    *_ec_public_key = pk;
+
+#else
+    EC_KEY *ec_key = NULL;
+    const EC_GROUP *gr;
+    const EC_POINT *pk;
+
+    ec_key = EVP_PKEY_get1_EC_KEY(cert_pub_key);
+    if (ec_key == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    gr = EC_KEY_get0_group(ec_key);
+
+    pk = EC_KEY_get0_public_key(ec_key);
+
+    *_ec_group = EC_GROUP_dup(gr);
+    if (*_ec_group == NULL) {
+        EC_KEY_free(ec_key);
+        ret = ENOMEM;
+        goto done;
+    }
+
+    *_ec_public_key = EC_POINT_dup(pk, gr);
+    EC_KEY_free(ec_key);
+    if (*_ec_public_key == NULL) {
+        EC_GROUP_free(*_ec_group);
+        ret = ENOMEM;
+        goto done;
+    }
+#endif
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
 static errno_t ec_pub_key_to_ssh(TALLOC_CTX *mem_ctx, EVP_PKEY *cert_pub_key,
                                  uint8_t **key_blob, size_t *key_size)
 {
     int ret;
     size_t c;
     uint8_t *buf = NULL;
     size_t buf_len;
-    EC_KEY *ec_key = NULL;
-    const EC_GROUP *ec_group = NULL;
-    const EC_POINT *ec_public_key = NULL;
+    EC_GROUP *ec_group = NULL;
+    EC_POINT *ec_public_key = NULL;
     BN_CTX *bn_ctx = NULL;
     int key_len;
     const char *identifier = NULL;
     int identifier_len;
     const char *header = NULL;
     int header_len;
 
-    ec_key = EVP_PKEY_get1_EC_KEY(cert_pub_key);
-    if (ec_key == NULL) {
-        ret = ENOMEM;
+    ret = sss_ec_get_key(bn_ctx, cert_pub_key, &ec_group, &ec_public_key);
+    if (ret != EOK) {
         goto done;
```

Comment:
Is it safe to do
```
    EC_GROUP_free(ec_group);
    EC_POINT_free(ec_public_key);
```
in the 'done:' section in this case?

Reply:
I changed to code so that it will (hopefully) always be safe to call it.